feat:add utc flag to date column

CHOIJEWON opened the pull request:
closes: #11515 

### Description of change

This PR adds UTC timezone support for date type columns by introducing a new utc option to ColumnOptions.

Currently, date columns always use local timezone when converting dates to strings, making it difficult to maintain consistent
timezone across different environments. This can cause issues in distributed systems or when the application server timezone differs from the database timezone.

With the new utc flag, developers can explicitly store and retrieve dates in UTC timezone, ensuring data consistency regardless of the server's local timezone setting.

### Example Usage:
```typescript
  @Entity()
  class Event {
      @PrimaryGeneratedColumn()
      id: number

      @Column({ type: "date" })
      localDate: Date  // Uses local timezone (existing behavior)

      @Column({ type: "date", utc: true })
      utcDate: Date  // Uses UTC timezone (new feature)
  }
```
### What changed
- Added utc?: boolean option to ColumnOptions interface
- Updated ColumnMetadata to store and propagate the UTC flag
- Modified DateUtils.mixedDateToDateString() to support UTC conversion
- Applied UTC flag handling across all database drivers (MySQL, PostgreSQL, SQLite, Oracle, SQL Server, Spanner, CockroachDB, SAP HANA, React Native)
- Added comprehensive test case verifying UTC vs local timezone behavior

### Backward Compatibility
✅ No breaking changes - utc defaults to false, maintaining existing local timezone behavior

### Pull-Request Checklist
-   [x] Code is up-to-date with the `master` branch
-   [x] This pull request links relevant issues as `Fixes #00000`
-   [x] There are new or updated unit tests validating the change
-   [x] Documentation has been updated to reflect this change

### Additional Notes

This is my first contribution to TypeORM and open source in general. I've done my best to follow the project's conventions and
guidelines, but I would greatly appreciate any feedback on:

- Code style and structure
- Test coverage and approach
- Documentation clarity
- Anything I might have missed or could improve

Thank you for taking the time to review this PR! I'm excited to contribute to TypeORM and look forward to learning from the
maintainers' feedback. 🙏

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

* **New Features**
  * Added UTC timezone support for date columns, allowing dates to be stored and retrieved in UTC rather than local timezone.
  * New `utc` configuration option available for date column types across all supported databases.

* **Documentation**
  * Updated documentation to describe the new UTC date column option and default behavior.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

gioboa reviewed (commented):
Thanks for your help @CHOIJEWON

alumni commented:
I'm thinking that the problem is a bit more complex - the server timezone is also important (and might be different than the one the app uses) - but I'm thinking we could let the users figure it out for now and fix it later.

I'm also not sure which database clients (maybe mssql?) already return a Date object (and also not sure how the client decides how to create the Date object).

In my case (SAP) I'm getting strings either in the server's TZ or in UTC, depending which function is called. TypeORM uses the non-UTC functions for create/update/delete date columns (so this option won't work with them). Telling it to use `CURRENT_UTCTIMESTAMP` instead of `CURRENT_TIMESTAMP` depending on this option might add to the complexity.

I would still recommend using UTC everywhere just to be safe, but at least this partially solves the problem on the application side :)

naorpeled reviewed (approved):
LGTM, left a nit comment, let me know what you think about it 🙏 
After that's addressed I think we'll be good to merge

CHOIJEWON commented on src/util/DateUtils.ts:
```diff
@@ -25,8 +25,20 @@ export class DateUtils {
-    static mixedDateToDateString(value: string | Date): string {
+    static mixedDateToDateString(
+        value: string | Date,
+        utc: boolean = false,
```
Sorry for the delayed reply — the day is just starting here in Korea 🙂
Your feedback is completely reasonable. I’ll apply the updates and push a new commit shortly.

CHOIJEWON commented:
@naorpeled 

Thank you for your valuable feedback! I've updated the implementation as you suggested.

### Changes Made:

I've refactored the `mixedDateToDateString` function parameter from a simple boolean to an object parameter to ensure future extensibility without breaking changes:

**src/util/DateUtils.ts:28-32**
```typescript
static mixedDateToDateString(
    value: string | Date,
    options?: { utc?: boolean },
): string {
    const utc = options?.utc ?? false
```

All database drivers (MySQL, PostgreSQL, SQLite, Oracle, SQL Server, Spanner, CockroachDB, SAP HANA, React Native) have been updated to use the new parameter format:

DateUtils.mixedDateToDateString(value, { utc: columnMetadata.utc })

This approach allows for potential future options to be added without requiring API changes.

Could you please review the updated implementation when you have a moment? I'd greatly appreciate any additional feedback.

Thank you!

gioboa reviewed (approved):
Thanks for your help @CHOIJEWON
